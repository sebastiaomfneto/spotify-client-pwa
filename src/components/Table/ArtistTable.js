import React from "react";
import { Link } from "react-router-dom";

import { Table } from "./Table";
import { FavoriteButton } from "../FavoriteButton";

import { buildImageSrc } from "../../helpers";

function buildGenresList(genres = []) {
  return genres.join(", ");
}

function buildPopularity(followers = 0) {
  switch (true) {
    case followers > 80:
      return "Hot";
    case followers >= 60 && followers <= 79:
      return "Cool";
    case followers >= 30 && followers <= 59:
      return "Regular";
    case followers < 30:
      return "Underground";
    default:
      return "";
  }
}

export function ArtistTable({ artists = [] }) {
  return (
    <Table columns={["", "Image", "Name", "Genres", "Popularity"]}>
      {artists.map((artist) => (
        <tr key={artist.id}>
          <td>
            <FavoriteButton id={artist.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(artist.images)}
              alt="artist"
              width="20px"
              height="20px"
            />
          </td>
          <td>
            <Link
              to={{
                pathname: `/artists/${artist.id}`,
                state: { artist },
              }}
            >
              {artist.name}
            </Link>
          </td>
          <td>{buildGenresList(artist.genres)}</td>
          <td>{buildPopularity(artist.followers.total)}</td>
        </tr>
      ))}
    </Table>
  );
}
