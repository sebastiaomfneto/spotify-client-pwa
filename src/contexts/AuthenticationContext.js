import React, { useState, useEffect } from "react";

import { getItem, setItem } from "../services/StorageService";

export const AuthenticationContext = React.createContext();

export function AuthenticationContextProvider({ children }) {
  const initialToken = getItem("token");
  const [token, setToken] = useState(initialToken);
  const [isAuthenticated, setIsAuthenticated] = useState(!!initialToken);

  useEffect(() => {
    setItem("token", token);
    setIsAuthenticated(!!token);
  }, [token]);

  return (
    <AuthenticationContext.Provider
      value={{
        token,
        isAuthenticated,
        setToken,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
