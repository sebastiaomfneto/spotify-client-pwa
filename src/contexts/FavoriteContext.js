import React, { useEffect, useReducer } from 'react';

import * as StorageService from '../services/StorageService';
import { NotImplementedError } from '../helpers';

export const FavoriteContext = React.createContext();

export const TOGGLE_FAVORITE = 'TOGGLE_FAVORITE';

function initializer(initialState) {
  return StorageService.getItem('favorites') ?? initialState;
}

function reducer(state, action) {
  switch (action.type) {
    case TOGGLE_FAVORITE:
      const isfavorited = state.some((f) => f === action.payload);
      if (isfavorited) {
        return state.filter((favorite) => favorite !== action.payload);
      }
      return state.concat(action.payload);
    default:
      throw new NotImplementedError();
  }
}

export function FavoriteContextProvider({ initialState = [], children }) {
  const [state, dispatch] = useReducer(reducer, initialState, initializer);

  useEffect(() => {
    StorageService.setItem('favorites', state);
  }, [state]);

  return (
    <FavoriteContext.Provider
      value={{
        state,
        dispatch,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
}
