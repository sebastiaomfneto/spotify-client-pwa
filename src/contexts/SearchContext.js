import React, { useReducer } from 'react';

import { NotImplementedError } from '../helpers';

export const SearchContext = React.createContext();

export const SET_SEARCH = 'SET_SEARCH';

const initialState = '';

function reducer(_state, action) {
  switch (action.type) {
    case SET_SEARCH:
      return action.payload;
    default:
      throw new NotImplementedError();
  }
}

export function SearchContextProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <SearchContext.Provider
      value={{
        state,
        dispatch,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
}
