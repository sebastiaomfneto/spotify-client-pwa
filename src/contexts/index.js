export * from './AuthenticationContext';
export * from './FavoriteContext';
export * from './SearchContext';
