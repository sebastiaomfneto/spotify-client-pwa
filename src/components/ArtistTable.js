import React from "react";
import { Link } from "react-router-dom";

import Table from "../components/Table";
import FavoriteButton from "../components/FavoriteButton";

import { buildImageSrc, buildGenresList, buildPopularity } from "../helpers";

export default function ArtistTable({ artists = [] }) {
  return (
    <Table columns={["", "Image", "Name", "Genres", "Popularity"]}>
      {artists.map((artist) => (
        <tr key={artist.id}>
          <td>
            <FavoriteButton id={artist.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(artist.images)}
              alt="artist"
              width="20px"
              height="20px"
            />
          </td>
          <td>
            <Link
              to={{
                pathname: `/artists/${artist.id}`,
                state: { artist },
              }}
            >
              {artist.name}
            </Link>
          </td>
          <td>{buildGenresList(artist.genres)}</td>
          <td>{buildPopularity(artist.followers.total)}</td>
        </tr>
      ))}
    </Table>
  );
}
