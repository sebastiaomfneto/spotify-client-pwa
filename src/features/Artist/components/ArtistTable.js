import React from 'react';
import { Link } from 'react-router-dom';

import { Table, FavoriteButton } from '../../../components';
import { buildImageSrc } from '../../../helpers';
import { buildPopularity } from '../helpers';

export function ArtistTable({ artists = [] }) {
  return (
    <Table columns={['', 'Image', 'Name', 'Genres', 'Popularity']}>
      {artists.map((artist) => (
        <tr key={artist.id}>
          <td>
            <FavoriteButton id={artist.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(artist.images)}
              alt="artist"
              width="20px"
              height="20px"
            />
          </td>
          <td>
            <Link
              to={{
                pathname: `/artists/${artist.id}`,
                state: { artist },
              }}
            >
              {artist.name}
            </Link>
          </td>
          <td>{artist.genres.join(', ')}</td>
          <td>{buildPopularity(artist.followers.total)}</td>
        </tr>
      ))}
    </Table>
  );
}
