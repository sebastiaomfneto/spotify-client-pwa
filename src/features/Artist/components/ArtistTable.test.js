import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';

import { FavoriteContext } from '../../../contexts/FavoriteContext';
import { ArtistTable } from './ArtistTable';

jest.mock('../../../services/StorageService');

const withProviders = (element, state = [], dispatch) => (
  <MemoryRouter>
    <FavoriteContext.Provider value={{ state, dispatch }}>
      {element}
    </FavoriteContext.Provider>
  </MemoryRouter>
);

describe('testing ArtistTable', () => {
  const artists = [
    {
      id: 1,
      name: 'TEST_ARTIST',
      images: [{ src: 'TEST_SRC' }],
      genres: ['TEST_GENRE_1', 'TEST_GENRE_2'],
      followers: { total: 1 },
    },
  ];

  test('should match with snapshot', () => {
    const { container } = render(
      withProviders(<ArtistTable artists={artists} />)
    );

    expect(container).toMatchSnapshot();
  });
});
