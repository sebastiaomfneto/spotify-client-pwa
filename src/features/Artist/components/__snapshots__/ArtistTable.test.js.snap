// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ArtistTable should match with snapshot 1`] = `
<div>
  <div
    class="Table"
  >
    <table>
      <thead>
        <tr>
          <th />
          <th>
            Image
          </th>
          <th>
            Name
          </th>
          <th>
            Genres
          </th>
          <th>
            Popularity
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <button
              class="FavoriteButton "
              data-testid="FavoriteButton-button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-heart fa-w-16 "
                data-icon="heart"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
          <td>
            <img
              alt="artist"
              height="20px"
              width="20px"
            />
          </td>
          <td>
            <a
              href="/artists/1"
            >
              TEST_ARTIST
            </a>
          </td>
          <td>
            TEST_GENRE_1, TEST_GENRE_2
          </td>
          <td>
            Underground
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
